torch/functional.py

```diff
@@ -803,39 +814,51 @@ def norm(input, p="fro", dim=None, keepdim=False, out=None, dtype=None):
+                return _VF.frobenius_norm(input)
+        if not isinstance(p, str):
+            return _VF.norm(input, p)
+
+    if isinstance(p, str):
         if p == "fro":
-            return torch._C._VariableFunctions.frobenius_norm(input)
-        elif p != "nuc":
-            return torch._C._VariableFunctions.norm(input, p)
-
-    if p == "fro":
-        if dtype is not None:
-            raise ValueError("dtype argument is not supported in frobenius norm")
-        if dim is None:
-            dim = tuple(range(ndim))
-        if out is None:
-            return torch._C._VariableFunctions.frobenius_norm(input, dim, keepdim=keepdim)
-        return torch._C._VariableFunctions.frobenius_norm(input, dim, keepdim=keepdim, out=out)
-    elif p == "nuc":
-        if dtype is not None:
-            raise ValueError("dtype argument is not supported in nuclear norm")
-        if dim is None:
+            if dtype is not None:
+                raise ValueError("dtype argument is not supported in frobenius norm")
+            if dim is None:
+                dim = [i for i in range(ndim)]  # noqa: C416 TODO: rewrite as list(range(m))
             if out is None:
-                return torch._C._VariableFunctions.nuclear_norm(input, keepdim=keepdim)
-            return torch._C._VariableFunctions.nuclear_norm(input, keepdim=keepdim, out=out)
-        return torch._C._VariableFunctions.nuclear_norm(input, dim, keepdim=keepdim, out=out)
+                return _VF.frobenius_norm(input, dim, keepdim=keepdim)
+            else:
+                return _VF.frobenius_norm(input, dim, keepdim=keepdim, out=out)
+        elif p == "nuc":
+            if dtype is not None:
+                raise ValueError("dtype argument is not supported in nuclear norm")
+            if dim is None:
+                if out is None:
+                    return _VF.nuclear_norm(input, keepdim=keepdim)
+                else:
+                    return _VF.nuclear_norm(input, keepdim=keepdim, out=out)
+            else:
+                if out is None:
+                    return _VF.nuclear_norm(input, dim, keepdim=keepdim)
+                else:
+                    return _VF.nuclear_norm(input, dim, keepdim=keepdim, out=out)
+        raise RuntimeError("only valid string values are 'fro' and 'nuc', found {}".format(p))
     else:
         if dim is None:
-            dim = tuple(range(ndim))
-        if out is None and dtype is None:
-            return torch._C._VariableFunctions.norm(input, p, dim, keepdim=keepdim)
-        elif out is None:
-            return torch._C._VariableFunctions.norm(input, p, dim, keepdim=keepdim, dtype=dtype)
-        elif dtype is None:
-            return torch._C._VariableFunctions.norm(input, p, dim, keepdim=keepdim, out=out)
-    return torch._C._VariableFunctions.norm(input, p, dim, keepdim=keepdim, dtype=dtype, out=out)
+            dim = [i for i in range(ndim)]  # noqa: C416 TODO: rewrite as list(range(m))
```

Comment:
I noticed this in at least one other preceding PR as well, perhaps we could consider adding this as a separate function to reduce duplicate code.

Reply:
Sounds good. I’m also going to try to add list(range()) support to TS in the next week